> @chuanqi129 could you help check whether the windows build failure is related to this PR

Hi @guangyey @cyyever I have checked the build log in https://github.com/intel/torch-xpu-ops/actions/runs/27247266536/artifacts/7553725529, it looks like related

```
[2512/3330] Building SYCL (Device) object torch_xpu_ops_sycl_common_kernels_gen_DistributionExponentialKernel.cpp.obj
FAILED: [code=1] caffe2/aten_xpu/src/CMakeFiles/torch_xpu_ops_sycl_common_kernels.dir/ATen/native/xpu/sycl/torch_xpu_ops_sycl_common_kernels_gen_DistributionExponentialKernel.cpp.obj C:/actions-runner/_work/torch-xpu-ops/pytorch/build/caffe2/aten_xpu/src/CMakeFiles/torch_xpu_ops_sycl_common_kernels.dir/ATen/native/xpu/sycl/torch_xpu_ops_sycl_common_kernels_gen_DistributionExponentialKernel.cpp.obj 
C:\Windows\system32\cmd.exe /C "cd /D C:\actions-runner\_work\torch-xpu-ops\pytorch\build\caffe2\aten_xpu\src\CMakeFiles\torch_xpu_ops_sycl_common_kernels.dir\ATen\native\xpu\sycl && C:\ProgramData\miniforge3\envs\windows_ci\Library\bin\cmake.exe -E make_directory C:/actions-runner/_work/torch-xpu-ops/pytorch/build/caffe2/aten_xpu/src/CMakeFiles/torch_xpu_ops_sycl_common_kernels.dir/ATen/native/xpu/sycl/. && C:\ProgramData\miniforge3\envs\windows_ci\Library\bin\cmake.exe -D verbose:BOOL=OFF -D generated_file:STRING=C:/actions-runner/_work/torch-xpu-ops/pytorch/build/caffe2/aten_xpu/src/CMakeFiles/torch_xpu_ops_sycl_common_kernels.dir/ATen/native/xpu/sycl/./torch_xpu_ops_sycl_common_kernels_gen_DistributionExponentialKernel.cpp.obj -P C:/actions-runner/_work/torch-xpu-ops/pytorch/build/caffe2/aten_xpu/src/CMakeFiles/torch_xpu_ops_sycl_common_kernels.dir/ATen/native/xpu/sycl/torch_xpu_ops_sycl_common_kernels_gen_DistributionExponentialKernel.cpp.obj.Release.cmake"
Intel(R) oneAPI DPC++/C++ Compiler for applications running on Intel(R) 64, Version 2026.0.0 Build 20260331
Copyright (C) 1985-2026 Intel Corporation. All rights reserved.

icx: warning: unknown argument ignored in clang-cl '-std=c++20'; did you mean '-Qstd=c++20'? [-Wunknown-argument]
icx: warning: overriding '-ffp-model=strict' option with '-ffp-contract=fast' [-Woverriding-option]
icx: warning: overriding '-ffp-model=strict' option with '-ffp-contract=fast' [-Woverriding-option]
warning: overriding currently unsupported rounding mode on this target [-Wunsupported-floating-point-opt]
warning: overriding currently unsupported use of floating point exceptions on this target [-Wunsupported-floating-point-opt]
In file included from C:\actions-runner\_work\torch-xpu-ops\pytorch\third_party\torch-xpu-ops\src\ATen\native\xpu\sycl\DistributionExponentialKernel.cpp:11:
In file included from C:\actions-runner\_work\torch-xpu-ops\pytorch\aten\src\ATen\ATen.h:7:
In file included from C:\actions-runner\_work\torch-xpu-ops\pytorch\aten\src\ATen\Context.h:4:
In file included from C:\actions-runner\_work\torch-xpu-ops\pytorch\aten\src\ATen\CPUGeneratorImpl.h:3:
In file included from C:\actions-runner\_work\torch-xpu-ops\pytorch\aten\src\ATen\core\Generator.h:18:
In file included from C:\actions-runner\_work\torch-xpu-ops\pytorch\c10\core\GeneratorImpl.h:8:
In file included from C:\actions-runner\_work\torch-xpu-ops\pytorch\c10\core\TensorImpl.h:8:
In file included from C:\actions-runner\_work\torch-xpu-ops\pytorch\c10\core\InferenceMode.h:3:
C:\actions-runner\_work\torch-xpu-ops\pytorch\c10\core\AutogradState.h(88,33): warning: default member initializer for bit-field is a C++20 extension [-Wc++20-extensions]
   88 |   bool view_replay_enabled_ : 1 = false;
      |                                 ^
C:\actions-runner\_work\torch-xpu-ops\pytorch\c10\core\AutogradState.h(89,45): warning: default member initializer for bit-field is a C++20 extension [-Wc++20-extensions]
   89 |   bool grad_layout_enforcement_enabled_ : 1 = true;
      |                                             ^
In file included from C:\actions-runner\_work\torch-xpu-ops\pytorch\third_party\torch-xpu-ops\src\ATen\native\xpu\sycl\DistributionExponentialKernel.cpp:14:
In file included from C:\actions-runner\_work\torch-xpu-ops\pytorch\third_party\torch-xpu-ops\src\ATen\native\xpu\sycl\DistributionTemplates.h:24:
In file included from C:\actions-runner\_work\torch-xpu-ops\pytorch\third_party\torch-xpu-ops\src\ATen\native\xpu\sycl\Philox4x32.h:48:
In file included from C:\actions-runner\_work\torch-xpu-ops\pytorch\aten\src\ATen\core\DistributionsHelper.h:3:
In file included from C:\actions-runner\_work\torch-xpu-ops\pytorch\aten\src\ATen\core\TransformationHelper.h:5:
C:\actions-runner\_work\torch-xpu-ops\pytorch\c10\util\MathConstants.h(15,30): error: no member named 'numbers' in namespace 'std'
   15 |   return static_cast<T>(std::numbers::e_v<double>);
      |                         ~~~~~^
C:\actions-runner\_work\torch-xpu-ops\pytorch\c10\util\MathConstants.h(15,49): error: expected '(' for function-style cast or type construction
   15 |   return static_cast<T>(std::numbers::e_v<double>);
      |                                           ~~~~~~^
C:\actions-runner\_work\torch-xpu-ops\pytorch\c10\util\MathConstants.h(15,50): error: expected expression
   15 |   return static_cast<T>(std::numbers::e_v<double>);
      |                                                  ^
C:\actions-runner\_work\torch-xpu-ops\pytorch\c10\util\MathConstants.h(20,30): error: no member named 'numbers' in namespace 'std'
   20 |   return static_cast<T>(std::numbers::egamma_v<double>);
      |                         ~~~~~^
C:\actions-runner\_work\torch-xpu-ops\pytorch\c10\util\MathConstants.h(20,54): error: expected '(' for function-style cast or type construction
   20 |   return static_cast<T>(std::numbers::egamma_v<double>);
      |                                                ~~~~~~^
C:\actions-runner\_work\torch-xpu-ops\pytorch\c10\util\MathConstants.h(20,55): error: expected expression
   20 |   return static_cast<T>(std::numbers::egamma_v<double>);
      |                                                       ^
C:\actions-runner\_work\torch-xpu-ops\pytorch\c10\util\MathConstants.h(25,30): error: no member named 'numbers' in namespace 'std'
   25 |   return static_cast<T>(std::numbers::inv_pi_v<double>);
      |                         ~~~~~^
C:\actions-runner\_work\torch-xpu-ops\pytorch\c10\util\MathConstants.h(25,54): error: expected '(' for function-style cast or type construction
   25 |   return static_cast<T>(std::numbers::inv_pi_v<double>);
      |                                                ~~~~~~^
C:\actions-runner\_work\torch-xpu-ops\pytorch\c10\util\MathConstants.h(25,55): error: expected expression
   25 |   return static_cast<T>(std::numbers::inv_pi_v<double>);
      |                                                       ^
C:\actions-runner\_work\torch-xpu-ops\pytorch\c10\util\MathConstants.h(30,30): error: no member named 'numbers' in namespace 'std'
   30 |   return static_cast<T>(std::numbers::inv_sqrtpi_v<double>);
      |                         ~~~~~^
C:\actions-runner\_work\torch-xpu-ops\pytorch\c10\util\MathConstants.h(30,58): error: expected '(' for function-style cast or type construction
   30 |   return static_cast<T>(std::numbers::inv_sqrtpi_v<double>);
      |                                                    ~~~~~~^
C:\actions-runner\_work\torch-xpu-ops\pytorch\c10\util\MathConstants.h(30,59): error: expected expression
   30 |   return static_cast<T>(std::numbers::inv_sqrtpi_v<double>);
      |                                                           ^
C:\actions-runner\_work\torch-xpu-ops\pytorch\c10\util\MathConstants.h(40,30): error: no member named 'numbers' in namespace 'std'
   40 |   return static_cast<T>(std::numbers::inv_sqrt3_v<double>);
      |                         ~~~~~^
C:\actions-runner\_work\torch-xpu-ops\pytorch\c10\util\MathConstants.h(40,57): error: expected '(' for function-style cast or type construction
   40 |   return static_cast<T>(std::numbers::inv_sqrt3_v<double>);
      |                                                   ~~~~~~^
C:\actions-runner\_work\torch-xpu-ops\pytorch\c10\util\MathConstants.h(40,58): error: expected expression
   40 |   return static_cast<T>(std::numbers::inv_sqrt3_v<double>);
      |                                                          ^
C:\actions-runner\_work\torch-xpu-ops\pytorch\c10\util\MathConstants.h(45,30): error: no member named 'numbers' in namespace 'std'
   45 |   return static_cast<T>(std::numbers::phi_v<double>);
      |                         ~~~~~^
C:\actions-runner\_work\torch-xpu-ops\pytorch\c10\util\MathConstants.h(45,51): error: expected '(' for function-style cast or type construction
   45 |   return static_cast<T>(std::numbers::phi_v<double>);
      |                                             ~~~~~~^
C:\actions-runner\_work\torch-xpu-ops\pytorch\c10\util\MathConstants.h(45,52): error: expected expression
   45 |   return static_cast<T>(std::numbers::phi_v<double>);
      |                                                    ^
C:\actions-runner\_work\torch-xpu-ops\pytorch\c10\util\MathConstants.h(50,30): error: no member named 'numbers' in namespace 'std'
   50 |   return static_cast<T>(std::numbers::ln10_v<double>);
      |                         ~~~~~^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
4 warnings and 20 errors generated.
CMake Error at torch_xpu_ops_sycl_common_kernels_gen_DistributionExponentialKernel.cpp.obj.Release.cmake:153 (message):
  Error generating file
  C:/actions-runner/_work/torch-xpu-ops/pytorch/build/caffe2/aten_xpu/src/CMakeFiles/torch_xpu_ops_sycl_common_kernels.dir/ATen/native/xpu/sycl/./torch_xpu_ops_sycl_common_kernels_gen_DistributionExponentialKernel.cpp.obj
```

--- a/cmake/BuildFlags.cmake
+++ b/cmake/BuildFlags.cmake
@@ -101,14 +101,17 @@ macro(set_build_flags)
     # gcc -shared host.o kernel.o device-code.o -o libxxx.so
     set(SYCL_KERNEL_OPTIONS ${SYCL_KERNEL_OPTIONS} -fno-sycl-unnamed-lambda)
     set(SYCL_KERNEL_OPTIONS ${SYCL_KERNEL_OPTIONS} -sycl-std=2020)
-    set(SYCL_KERNEL_OPTIONS ${SYCL_KERNEL_OPTIONS} -std=${CPP_STD})
     if(CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")
+      # icx runs in clang-cl mode on Windows and silently ignores GCC-style
+      # -std=; use the MSVC-style spelling so device code is really C++20.
+      set(SYCL_KERNEL_OPTIONS ${SYCL_KERNEL_OPTIONS} -Qstd=${CPP_STD})
       set(SYCL_KERNEL_OPTIONS ${SYCL_KERNEL_OPTIONS} /fp:strict)
       set(SYCL_KERNEL_OPTIONS ${SYCL_KERNEL_OPTIONS} /Qfma)
       set(SYCL_KERNEL_OPTIONS ${SYCL_KERNEL_OPTIONS} /Qftz-)
       # Suppress warnings about dllexport.
       set(SYCL_KERNEL_OPTIONS ${SYCL_KERNEL_OPTIONS} -Wno-ignored-attributes)
     elseif(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+      set(SYCL_KERNEL_OPTIONS ${SYCL_KERNEL_OPTIONS} -std=${CPP_STD})
       set(SYCL_KERNEL_OPTIONS ${SYCL_KERNEL_OPTIONS} -Wno-absolute-value)
       set(SYCL_KERNEL_OPTIONS ${SYCL_KERNEL_OPTIONS} -fno-fast-math)
       # -fma which we used before is an alias used for -ffp-contract=fast for compatibility reasons

--- a/cmake/BuildFlags.cmake
+++ b/cmake/BuildFlags.cmake
@@ -102,8 +102,8 @@ macro(set_build_flags)
     set(SYCL_KERNEL_OPTIONS ${SYCL_KERNEL_OPTIONS} -fno-sycl-unnamed-lambda)
     set(SYCL_KERNEL_OPTIONS ${SYCL_KERNEL_OPTIONS} -sycl-std=2020)
     if(CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")
-      # icx runs in clang-cl mode on Windows and silently ignores GCC-style
-      # -std=; use the MSVC-style spelling so device code is really C++20.
+      # On Windows icx uses the clang-cl driver, which ignores -std= with
+      # only a warning; spell it as -Qstd= so device code is really C++20.
       set(SYCL_KERNEL_OPTIONS ${SYCL_KERNEL_OPTIONS} -Qstd=${CPP_STD})
       set(SYCL_KERNEL_OPTIONS ${SYCL_KERNEL_OPTIONS} /fp:strict)
       set(SYCL_KERNEL_OPTIONS ${SYCL_KERNEL_OPTIONS} /Qfma)

--- a/src/ATen/native/xpu/sycl/Reduce.h
+++ b/src/ATen/native/xpu/sycl/Reduce.h
@@ -202,7 +202,18 @@ inline int last_pow2(int n) {
 }
 
 static void reduce_fraction(size_t& numerator, size_t& denominator) {
+#ifdef _MSC_VER
+  // MSVC's std::gcd does runtime ISA dispatch through a global variable and
+  // CPU builtins, which cannot be used in device code; use Euclid's algorithm.
+  size_t gcd = denominator;
+  size_t b = numerator;
+  while (b != 0) {
+    gcd %= b;
+    std::swap(gcd, b);
+  }
+#else
   auto gcd = std::gcd(numerator, denominator);
+#endif
   numerator /= gcd;
   denominator /= gcd;
 }

